src/style/css_syntax/mod.rs

```diff
@@ -0,0 +1,289 @@
+            "overflow" => *overflow = parse!(),
+            "overflow-x" => overflow.x = parse!(),
+            "overflow-y" => overflow.y = parse!(),
+            // https://w3c.github.io/csswg-drafts/css-position-3/#position-property
+            "position" => *position = parse!(),
+            // https://w3c.github.io/csswg-drafts/css-position-3/#inset-shorthands
+            "inset" => *inset = parse!(),
+            "top" => inset.top = parse!(),
+            "right" => inset.right = parse!(),
+            "bottom" => inset.bottom = parse!(),
+            "left" => inset.left = parse!(),
+            // https://drafts.csswg.org/css-sizing/#preferred-size-properties
+            "width" => size.width = non_negative(parse!()),
+            "height" => size.height = non_negative(parse!()),
+            // https://drafts.csswg.org/css-sizing/#min-size-properties
+            "min-width" => min_size.width = non_negative(parse!()),
+            "min-height" => min_size.height = non_negative(parse!()),
+            // https://drafts.csswg.org/css-sizing/#max-size-properties
+            "max-width" => max_size.width = non_negative(parse!()),
+            "max-height" => max_size.height = non_negative(parse!()),
+            // https://w3c.github.io/csswg-drafts/css-sizing-4/#aspect-ratio
+            "aspect-ratio" => *aspect_ratio = MaybeAuto::to_opt_f32(parse!()),
+            // https://drafts.csswg.org/css2/#margin-properties
+            "margin" => *margin = parse!(),
+            "margin-top" => margin.top = parse!(),
+            "margin-right" => margin.right = parse!(),
+            "margin-bottom" => margin.bottom = parse!(),
+            "margin-left" => margin.left = parse!(),
+            // https://drafts.csswg.org/css2/#padding-properties
+            "padding" => *padding = non_negative(parse!()),
+            "padding-top" => padding.top = non_negative(parse!()),
+            "padding-right" => padding.right = non_negative(parse!()),
+            "padding-bottom" => padding.bottom = non_negative(parse!()),
+            "padding-left" => padding.left = non_negative(parse!()),
+
+            _ => {
+                return Err(input.new_custom_error(ParseErrorKind::InvalidOrUnknownProperty))
+            }
+        }
+        Ok(())
+    }
+}
+
+// Default methods always return Err for no supported at-rule
+impl<'i> cssparser::AtRuleParser<'i> for DeclarationParser<'_> {
+    type Prelude = ();
+    type AtRule = ();
+    type Error = ParseErrorKind<'i>;
+}
+
+// Default methods always return Err for no supported nested qualified rule
+impl<'i> cssparser::QualifiedRuleParser<'i> for DeclarationParser<'_> {
+    type Prelude = ();
+    type QualifiedRule = ();
+    type Error = ParseErrorKind<'i>;
+}
+
+trait CssValue: Sized {
+    fn parse<'i, 't>(input: &mut Parser<'i, 't>, taffy: &Taffy) -> Result<Self, ParseError<'i>>;
+
+    fn serialize(&self, dest: &mut String, taffy: &Taffy);
+}
```

Comment:
I'm thinking it might make sense to split this up into two traits (one for parsing, one for serializing). That way we can put parsing and serializing code behind separate feature flags if we want to.

Reply:
I'm also wondering if it would make sense to just make these the `Display` and `FromStr` traits. Do you have any insight into the pros and cons of using the standard traits vs. custom ones?

Reply:
I’m not super convinced of the value of enabling one of parsing or serializing without the other but if you prefer yes it could be two traits.

Parsing needs a signature different than `FromStr`. For parsing a value of various Rust types while parsing a declaration, we want to keep using the same tokenizer and `cssparser::Parser`. Finding the right start and end to extract a `&str` slice and create another tokenizer for that would be extra work, and would lose the ability for errors to carry accurate source location (line numbers).

If a `&Taffy` parameter is no longer needed (see above) then serialization could have the same signature as `Display` but I think it still represents different intent. Values that can contain arbitrary text may need backslash-escapes for serialization to CSS syntax while Display (for presentation to users of the program) probably doesn’t want that.